crates/ruma-macros/src/api/api_metadata.rs

```diff
@@ -303,13 +290,126 @@ impl Parse for FieldValue {
     }
 }
 
-#[derive(Clone)]
+#[derive(Debug, PartialEq)]
+pub struct History {
+    pub(super) entries: Vec<HistoryEntry>,
+    misc: MiscVersioning,
+}
+
+impl History {
+    // TODO(j0j0): remove after codebase conversion is complete
+    /// Construct a History object from legacy parts.
+    pub fn construct(
+        deprecated: Option<MatrixVersionLiteral>,
+        removed: Option<MatrixVersionLiteral>,
+        unstable_path: Option<EndpointPath>,
+        r0_path: Option<EndpointPath>,
+        stable_path_and_version: Option<(EndpointPath, MatrixVersionLiteral)>,
+    ) -> Self {
+        // Unfortunately can't `use` associated constants
+        const V1_0: MatrixVersionLiteral = MatrixVersionLiteral::V1_0;
+
+        let unstable = unstable_path.map(|path| HistoryEntry::Unstable { path });
+        let r0 = r0_path.map(|path| HistoryEntry::Stable { path, version: V1_0 });
+        let stable = stable_path_and_version.map(|(path, mut version)| {
+            if r0.is_some() && version == V1_0 {
+                version = MatrixVersionLiteral::V1_1;
+            }
```

Comment:
I'm not sure I understand why this is necessary. Aren't v1.0 endpoints always using the r0 path?

And why change the matrix version?

Reply:
I don't know either, either this is @ShadowJonathan's change, or me screwing up a rebase / the cleanup.

Reply:
The `version` here comes from the `added` field, this is a special case because we have;
- `added` (which says v1.0)
- `r0_path`
- `stable_path`

`r0_path` is already added under v1.0 implicitly, and we should have a check somewhere that makes sure `added` doesn't say something else in this situation, so the `stable_path` in this instance becomes v1.1, as then it's documenting the `/v3/` paths that came with the move to v1.1.

Reply:
Right I see it now.